app/Model/CoIdentifierAssignment.php

```diff
@@ -499,7 +502,14 @@ private function checkInsert($coIdentifierAssignment, $objType, $assignEmail, $o
     
     $coId = $coIdentifierAssignment['CoIdentifierAssignment']['co_id'];
     $fk = Inflector::underscore($objType) . "_id";
-    
+
+
+    // Begin a transaction. This is more because we need to ensure the integrity of
+    // data between SELECT and INSERT/UPDATE than that we expect to rollback.
+
+    $dbc = $this->getDataSource();
```

Comment:
Why are we starting a new transaction here? `checkInsert()` should be called from within a transaction.

Reply:
@boshrin thank you for the clarification. `checkInsert()` is using `$dbc->rollback()` but there is no initialization of the `$dbc`. I wrongly assumed that the function should handle the transaction by itself. Following your comment, i removed the transaction initialization and passed the `$dbc` variable as a parameter. Now `checkInsert()` has a reference to the transaction and can call a rollback if an exception occurs.

Reply:
`checkInsert()` doesn't need to call the rollback. If it throws an Exception, the parent code will catch it and call `rollback()`.

Reply:
Done